File: src/main/scala/dregex/impl/Compiler.scala
package dregex.impl

import dregex.InvalidRegexException

import java.util.stream.Collectors
import scala.collection.mutable.Buffer
import scala.jdk.CollectionConverters._

/**
  * Take a regex AST and produce a NFA.
  * Except when noted the Thompson-McNaughton-Yamada algorithm is used.
  * Reference: http://stackoverflow.com/questions/11819185/steps-to-creating-an-nfa-from-a-regular-expression
  */
class Compiler(intervalMapping: java.util.Map[RegexTree.AbstractRange, java.util.List[CharInterval]]) {

  import RegexTree._

  /**
    * Transform a regular expression abstract syntax tree into a corresponding NFA
    */
  def fromTree(ast: Node): Dfa[SimpleState] = {
    val initial = new SimpleState
    val accepting = new SimpleState
    val transitions = fromTreeImpl(ast, initial, accepting)
    val nfa = new Nfa(initial, transitions.asJava, Set[State](accepting).asJava)
    DfaAlgorithms.rewriteWithSimpleStates(DfaAlgorithms.fromNfa(nfa))
  }

  private def fromTreeImpl(node: Node, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    node match {

      // base case

      case range: AbstractRange =>
        val intervals = intervalMapping.get(range)
        intervals.stream().map(interval => new Nfa.Transition(from, to, interval)).collect(Collectors.toList()).asScala.toSeq

      // recurse

      case CharSet(intervals) =>
        fromTreeImpl(Disj(intervals), from, to)

      case juxt: Juxt =>
        processJuxt(combineNegLookaheads(juxt), from, to)

      case la: Lookaround =>
        fromTreeImpl(Juxt(Seq(la)), from, to)

      case disj: Disj =>
        processDisj(disj, from, to)

      case rep: Rep =>
        processRep(rep, from, to)

      case Intersection(left, right) =>
        processOp(DfaAlgorithms.intersect, left, right, from, to)

      case Union(left, right) =>
        processOp(DfaAlgorithms.union, left, right, from, to)

      case Difference(left, right) =>
        processOp(DfaAlgorithms.diff, left, right, from, to)

      case cg: PositionalCaptureGroup =>
        processCaptureGroup(cg.value, from, to)

      case _: NamedCaptureGroup =>
        throw new InvalidRegexException("named capture groups are not supported")
    }
  }

  /**
    * Lookaround constructions are transformed in equivalent DFA operations, and the result of those trivially transformed
    * into a NFA again for insertion into the outer expression.
    *
    * (?=B)C is transformed into C ∩ B.*
    * (?!B)C is transformed into C - B.*
    * A(?<=B) is transformed into A ∩ .*B
    * A(?<!B) is transformed into A - .*B
    *
    * In the case of more than one lookaround, the transformation is applied recursively.
    *
    * *
    * NOTE: Only lookahead is currently implemented
    */
  private def processJuxt(juxt: Juxt, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    import Direction._
    import Condition._
    findLookaround(juxt.values) match {
      case Some(i) =>
        val prefix = juxt.values.slice(0, i)
        val suffix = juxt.values.slice(i + 1, juxt.values.size)
        val wildcard = Rep(min = 0, max = None, value = Wildcard)
        juxt.values(i).asInstanceOf[Lookaround] match {
          case Lookaround(Ahead, cond, value) =>
            val rightSide: Node = cond match {
              case Positive => Intersection(Juxt(suffix), Juxt(Seq(value, wildcard)))
              case Negative => Difference(Juxt(suffix), Juxt(Seq(value, wildcard)))
            }
            if (prefix.isEmpty)
              fromTreeImpl(rightSide, from, to)
            else
              fromTreeImpl(Juxt(prefix :+ rightSide), from, to)
          case Lookaround(Behind, cond, value) =>
            val leftSide: Node = cond match {
              case Positive => Intersection(Juxt(prefix), Juxt(Seq(value, wildcard)))
              case Negative => Difference(Juxt(prefix), Juxt(Seq(value, wildcard)))
            }
            if (suffix.isEmpty)
              fromTreeImpl(leftSide, from, to)
            else
              fromTreeImpl(Juxt(leftSide +: suffix), from, to)
        }
      case None =>
        processJuxtNoLookaround(juxt, from, to)
    }
  }

  def findLookaround(args: Seq[Node]): Option[Int] = {
    val found = args.zipWithIndex.find { case (x, i) => x.isInstanceOf[Lookaround] }
    found.map { case (_, idx) => idx }
  }

  /**
    * Optimization: combination of consecutive negative lookahead constructions
    * (?!a)(?!b)(?!c) gets combined to (?!a|b|c), which is faster to process.
    * This optimization should be applied before the look-around's are expanded to intersections and differences.
    */
  def combineNegLookaheads(juxt: Juxt): Juxt = {
    import Direction._
    import Condition._
    val newValues = juxt.values.foldLeft(Seq[Node]()) { (acc, x) =>
      (acc, x) match {
        case (init :+ Lookaround(Ahead, Negative, v1), Lookaround(Ahead, Negative, v2)) =>
          init :+ Lookaround(Ahead, Negative, Disj(Seq(v1, v2)))
        case _ =>
          acc :+ x
      }
    }
    Juxt(newValues)
  }

  private def processJuxtNoLookaround(juxt: Juxt, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    juxt match {
      case Juxt(Seq()) =>
        Seq(new Nfa.Transition(from, to, Epsilon.instance))

      case Juxt(Seq(head)) =>
        fromTreeImpl(head, from, to)

      case Juxt(init :+ last) =>
        // doing this iteratively prevents stack overflows in the case of long literal strings
        val transitions = Buffer[Nfa.Transition]()
        var prev = from
        for (part <- init) {
          val int = new SimpleState
          transitions ++= fromTreeImpl(part, prev, int)
          prev = int
        }
        transitions ++= fromTreeImpl(last, prev, to)
        transitions.to(Seq)
    }
  }

  private def processDisj(disj: Disj, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    disj match {
      case Disj(Seq()) =>
        Seq()
      case Disj(parts) =>
        parts.map(part => fromTreeImpl(part, from, to)).flatten
    }
  }

  private def processRep(rep: Rep, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    rep match {

      // trivial cases

      case Rep(1, Some(1), value) =>
        fromTreeImpl(value, from, to)

      case Rep(0, Some(0), value) =>
        Seq(new Nfa.Transition(from, to, Epsilon.instance))

      // infinite repetitions

      case Rep(n, None, value) if n > 1 =>
        val juxt = Juxt(Seq.fill(n)(value) :+ Rep(0, None, value))
        fromTreeImpl(juxt, from, to)

      case Rep(1, None, value) =>
        val int1 = new SimpleState
        val int2 = new SimpleState
        fromTreeImpl(value, int1, int2) :+
          new Nfa.Transition(from, int1, Epsilon.instance) :+
          new Nfa.Transition(int2, to, Epsilon.instance) :+
          new Nfa.Transition(int2, int1, Epsilon.instance)

      case Rep(0, None, value) =>
        val int1 = new SimpleState
        val int2 = new SimpleState
        fromTreeImpl(value, int1, int2) :+
          new Nfa.Transition(from, int1, Epsilon.instance) :+
          new Nfa.Transition(int2, to, Epsilon.instance) :+
          new Nfa.Transition(from, to, Epsilon.instance) :+
          new Nfa.Transition(int2, int1, Epsilon.instance)

      // finite repetitions

      case Rep(n, Some(m), value) if n > 1 =>
        val x = n - 1
        val juxt = Juxt(Seq.fill(x)(value) :+ Rep(1, Some(m - x), value))
        fromTreeImpl(juxt, from, to)

      case Rep(1, Some(m), value) if m > 0 =>
        // doing this iteratively prevents stack overflows in the case of long repetitions
        val int1 = new SimpleState
        val transitions = Buffer[Nfa.Transition]()
        transitions ++= fromTreeImpl(value, from, int1)
        var prev = int1
        for (i <- 1 until m - 1) {
          val int = new SimpleState
          transitions ++= fromTreeImpl(value, prev, int)
          transitions += new Nfa.Transition(prev, to, Epsilon.instance)
          prev = int
        }
        transitions ++= fromTreeImpl(value, prev, to)
        transitions += new Nfa.Transition(prev, to, Epsilon.instance)
        transitions.to(Seq)

      case Rep(0, Some(m), value) if m > 0 =>
        // doing this iteratively prevents stack overflows in the case of long repetitions
        val transitions = Buffer[Nfa.Transition]()
        var prev = from
        for (i <- 0 until m - 1) {
          val int = new SimpleState
          transitions ++= fromTreeImpl(value, prev, int)
          transitions += new Nfa.Transition(prev, to, Epsilon.instance)
          prev = int
        }
        transitions ++= fromTreeImpl(value, prev, to)
        transitions += new Nfa.Transition(prev, to, Epsilon.instance)
        transitions.to(Seq)

    }
  }

  private def processOp(
      operation: DfaAlgorithms.BinaryOp[SimpleState],
      left: Node,
      right: Node,
      from: SimpleState,
      to: SimpleState): Seq[Nfa.Transition] = {
    val leftDfa = fromTree(left)
    val rightDfa = fromTree(right)
    val result = DfaAlgorithms.toNfa(operation(leftDfa, rightDfa))
    result.transitions.asScala ++
      result.accepting.asScala.to(Seq).map(acc => new Nfa.Transition(acc, to, Epsilon.instance)) ++
      Seq(new Nfa.Transition(from, result.initial, Epsilon.instance))
  }.toSeq

  def processCaptureGroup(value: Node, from: SimpleState, to: SimpleState): Seq[Nfa.Transition] = {
    val int1 = new SimpleState
    val int2 = new SimpleState
    fromTreeImpl(value, int1, int2) :+
      new Nfa.Transition(from, int1, Epsilon.instance) :+
      new Nfa.Transition(int2, to, Epsilon.instance)
  }

}

File: src/main/java/dregex/impl/Compiler.java
package dregex.impl;

import dregex.InvalidRegexException;
import dregex.impl.tree.*;
import java.time.Duration;
import java.util.*;
import java.util.function.BiFunction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Take a regex AST and produce a NFA.
 * Except when noted the Thompson-McNaughton-Yamada algorithm is used.
 * <a href="http://stackoverflow.com/questions/11819185/steps-to-creating-an-nfa-from-a-regular-expression">Reference</a>
 */
public class Compiler {

    private static final Logger logger = LoggerFactory.getLogger(Compiler.class);

    private final Map<AbstractRange, List<CharInterval>> intervalMapping;

    public Compiler(Map<AbstractRange, List<CharInterval>> intervalMapping) {
        this.intervalMapping = Map.copyOf(intervalMapping);
    }

    /**
     * Transform a regular expression abstract syntax tree into a corresponding DFA
     */
    public Dfa fromTree(Node ast) {
        var start = System.nanoTime();
        var initial = new SimpleState();
        var accepting = new SimpleState();
        List<Nfa.Transition> transitions = new ArrayList<>();
        addTransitionsFromNode(transitions, ast, initial, accepting);
        var nfa = new Nfa(initial, transitions, Set.of(accepting));
        var dfa = DfaAlgorithms.rewriteWithSimpleStates(DfaAlgorithms.fromNfa(nfa));
        var time = Duration.ofNanos(System.nanoTime() - start);
        logger.trace("DFA compiled in {}", time);
        return dfa;
    }

    private void addTransitionsFromNode(List<Nfa.Transition> transitions, Node node, SimpleState from, SimpleState to) {
        if (node instanceof AbstractRange) {
            // base case
            var range = (AbstractRange) node;
            for (var interval : intervalMapping.get(range)) {
                transitions.add(new Nfa.Transition(from, to, interval));
            }
        } else if (node instanceof CharSet) {
            var set = (CharSet) node;
            addTransitionsFromNode(transitions, new Disj(set.ranges), from, to);
        } else if (node instanceof Juxt) {
            // this optimization should be applied before the lookarounds are expanded to intersections and differences
            var juxt = (Juxt) node;
            addTransitionsFromJuxt(transitions, CompilerHelper.combineNegLookaheads(juxt), from, to);
        } else if (node instanceof Lookaround) {
            var la = (Lookaround) node;
            addTransitionsFromNode(transitions, new Juxt(la), from, to);
        } else if (node instanceof Disj) {
            var disj = (Disj) node;
            addTransitionsFromDisj(transitions, disj, from, to);
        } else if (node instanceof Rep) {
            var rep = (Rep) node;
            addTransitionsFromRep(transitions, rep, from, to);
        } else if (node instanceof Intersection) {
            var intersection = (Intersection) node;
            addTransitionsFromOperation(
                    transitions, DfaAlgorithms::doIntersect, intersection.left, intersection.right, from, to);
        } else if (node instanceof Union) {
            var union = (Union) node;
            addTransitionsFromOperation(transitions, DfaAlgorithms::union, union.left, union.right, from, to);
        } else if (node instanceof Difference) {
            var difference = (Difference) node;
            addTransitionsFromOperation(transitions, DfaAlgorithms::diff, difference.left, difference.right, from, to);
        } else if (node instanceof PositionalCaptureGroup) {
            var cg = (PositionalCaptureGroup) node;
            addTransitionsFromCaptureGroup(transitions, cg.value, from, to);
        } else if (node instanceof NamedCaptureGroup) {
            throw new InvalidRegexException("named capture groups are not supported");
        } else {
            throw new IllegalArgumentException();
        }
    }

    /**
     * Lookaround constructions are transformed in equivalent DFA operations, and the result of those trivially transformed
     * into a NFA again for insertion into the outer expression.
     * <p>
     * (?=B)C is transformed into C ∩ B.*
     * (?!B)C is transformed into C - B.*
     * A(?<=B) is transformed into A ∩ .*B
     * A(?<!B) is transformed into A - .*B
     * <p>
     * In the case of more than one lookaround, the transformation is applied recursively.
     * <p>
     * *
     * NOTE: Only lookahead is currently implemented
     */
    void addTransitionsFromJuxt(List<Nfa.Transition> transitions, Juxt juxt, SimpleState from, SimpleState to) {
        var lookaroundIdx = CompilerHelper.findLookaround(juxt.values);
        if (lookaroundIdx != -1) {
            var prefix = juxt.values.subList(0, lookaroundIdx);
            var suffix = juxt.values.subList(lookaroundIdx + 1, juxt.values.size());
            var wildcard = new Rep(0, Optional.empty(), Wildcard.instance);
            var lookaround = (Lookaround) juxt.values.get(lookaroundIdx);
            switch (lookaround.dir) {
                case Ahead:
                    Operation rightSide;
                    switch (lookaround.cond) {
                        case Positive:
                            rightSide = new Intersection(new Juxt(suffix), new Juxt(lookaround.value, wildcard));
                            break;
                        case Negative:
                            rightSide = new Difference(new Juxt(suffix), new Juxt(lookaround.value, wildcard));
                            break;
                        default:
                            throw new IllegalStateException();
                    }
                    if (prefix.isEmpty()) {
                        addTransitionsFromNode(transitions, rightSide, from, to);
                    } else {
                        List<Node> nodes = new ArrayList<>(prefix.size() + 1);
                        nodes.addAll(prefix);
                        nodes.add(rightSide);
                        addTransitionsFromNode(transitions, new Juxt(nodes), from, to);
                    }
                    break;
                case Behind:
                    Operation leftSide;
                    switch (lookaround.cond) {
                        case Positive:
                            leftSide = new Intersection(new Juxt(prefix), new Juxt(lookaround.value, wildcard));
                            break;
                        case Negative:
                            leftSide = new Difference(new Juxt(prefix), new Juxt(lookaround.value, wildcard));
                            break;
                        default:
                            throw new IllegalStateException();
                    }
                    if (suffix.isEmpty()) {
                        addTransitionsFromNode(transitions, leftSide, from, to);
                    } else {
                        List<Node> nodes = new ArrayList<>(suffix.size() + 1);
                        nodes.add(leftSide);
                        nodes.addAll(suffix);
                        addTransitionsFromNode(transitions, new Juxt(nodes), from, to);
                    }
                    break;
                default:
                    throw new IllegalStateException();
            }
        } else {
            addTransitionsFromJuxtNoLookaround(transitions, juxt, from, to);
        }
    }

    public void addTransitionsFromJuxtNoLookaround(
            List<Nfa.Transition> transitions, Juxt juxt, SimpleState from, SimpleState to) {
        if (juxt.values.isEmpty()) {
            transitions.add(new Nfa.Transition(from, to, Epsilon.instance));
        } else if (juxt.values.size() == 1) {
            addTransitionsFromNode(transitions, juxt.values.get(0), from, to);
        } else {
            // doing this iteratively prevents stack overflows in the case of long literal strings
            var prev = from;
            for (var part : juxt.values.subList(0, juxt.values.size() - 1)) {
                var intermediate = new SimpleState();
                addTransitionsFromNode(transitions, part, prev, intermediate);
                prev = intermediate;
            }
            addTransitionsFromNode(transitions, juxt.values.get(juxt.values.size() - 1), prev, to);
        }
    }

    private void addTransitionsFromOperation(
            List<Nfa.Transition> transitions,
            BiFunction<Dfa, Dfa, Dfa> operation,
            Node left,
            Node right,
            SimpleState from,
            SimpleState to) {
        var leftDfa = fromTree(left);
        var rightDfa = fromTree(right);
        var result = DfaAlgorithms.toNfa(operation.apply(leftDfa, rightDfa));
        transitions.addAll(result.transitions);
        for (var acc : result.accepting) {
            transitions.add(new Nfa.Transition(acc, to, Epsilon.instance));
        }
        transitions.add(new Nfa.Transition(from, result.initial, Epsilon.instance));
    }

    private void addTransitionsFromCaptureGroup(
            List<Nfa.Transition> transitions, Node value, SimpleState from, SimpleState to) {
        var int1 = new SimpleState();
        var int2 = new SimpleState();
        transitions.add(new Nfa.Transition(from, int1, Epsilon.instance));
        transitions.add(new Nfa.Transition(int2, to, Epsilon.instance));
        addTransitionsFromNode(transitions, value, int1, int2);
    }

    private void addTransitionsFromRep(List<Nfa.Transition> transitions, Rep rep, SimpleState from, SimpleState to) {
        if (rep.min == 1 && rep.max.isPresent() && rep.max.get() == 1) {

            addTransitionsFromNode(transitions, rep.value, from, to);

        } else if (rep.min == 0 && rep.max.isPresent() && rep.max.get() == 0) {

            transitions.add(new Nfa.Transition(from, to, Epsilon.instance));

        } else if (rep.min > 1 && rep.max.isEmpty()) {

            List<Node> juxtValues = new ArrayList<>(rep.min + 1);
            juxtValues.addAll(Collections.nCopies(rep.min, rep.value));
            juxtValues.add(new Rep(0, Optional.empty(), rep.value));
            addTransitionsFromNode(transitions, new Juxt(juxtValues), from, to);

        } else if (rep.min == 1 && rep.max.isEmpty()) {

            var int1 = new SimpleState();
            var int2 = new SimpleState();
            transitions.add(new Nfa.Transition(from, int1, Epsilon.instance));
            transitions.add(new Nfa.Transition(int2, to, Epsilon.instance));
            transitions.add(new Nfa.Transition(int2, int1, Epsilon.instance));
            addTransitionsFromNode(transitions, rep.value, int1, int2);

        } else if (rep.min == 0 && rep.max.isEmpty()) {

            var int1 = new SimpleState();
            var int2 = new SimpleState();
            transitions.add(new Nfa.Transition(from, int1, Epsilon.instance));
            transitions.add(new Nfa.Transition(int2, to, Epsilon.instance));
            transitions.add(new Nfa.Transition(from, to, Epsilon.instance));
            transitions.add(new Nfa.Transition(int2, int1, Epsilon.instance));
            addTransitionsFromNode(transitions, rep.value, int1, int2);

        } else if (rep.min > 1 && rep.max.isPresent()) {

            int x = rep.min - 1;
            List<Node> juxtValues = new ArrayList<>(x + 1);
            juxtValues.addAll(Collections.nCopies(x, rep.value));
            juxtValues.add(new Rep(1, Optional.of(rep.max.get() - x), rep.value));
            addTransitionsFromNode(transitions, new Juxt(juxtValues), from, to);

        } else if (rep.min == 1 && rep.max.isPresent() && rep.max.get() > 0) {

            // doing this iteratively prevents stack overflows in the case of long repetitions
            var int1 = new SimpleState();
            addTransitionsFromNode(transitions, rep.value, from, int1);
            var prev = int1;
            for (int i = 1; i < rep.max.get() - 1; i++) {
                var intermediate = new SimpleState();
                transitions.add(new Nfa.Transition(prev, to, Epsilon.instance));
                addTransitionsFromNode(transitions, rep.value, prev, intermediate);
                prev = intermediate;
            }
            transitions.add(new Nfa.Transition(prev, to, Epsilon.instance));
            addTransitionsFromNode(transitions, rep.value, prev, to);

        } else if (rep.min == 0 && rep.max.isPresent() && rep.max.get() > 0) {

            // doing this iteratively prevents stack overflows in the case of long repetitions
            var prev = from;
            for (int i = 0; i < rep.max.get() - 1; i++) {
                var intermediate = new SimpleState();
                transitions.add(new Nfa.Transition(prev, to, Epsilon.instance));
                addTransitionsFromNode(transitions, rep.value, prev, intermediate);
                prev = intermediate;
            }
            transitions.add(new Nfa.Transition(prev, to, Epsilon.instance));
            addTransitionsFromNode(transitions, rep.value, prev, to);

        } else {
            throw new IllegalArgumentException();
        }
    }

    private void addTransitionsFromDisj(List<Nfa.Transition> transitions, Disj disj, SimpleState from, SimpleState to) {
        for (Node part : disj.values) {
            addTransitionsFromNode(transitions, part, from, to);
        }
    }
}
